### DISCUSSION TITLE
Introducing cmake-format

### DISCUSSION THREAD

**Discussion by OlivierLDff** [KDAB/cxx-qt#505] [Ideas]
C++ code is formatted by clang format, rust with rustfmt, cargo file with magic (I don't know what format toml, but it is auto formatted somehow in my ide). Only cmake file are manually formatted which is annoying.

Basic config `.cmake-format` could look like that to match the style already used.

```yaml
format:
  tab_size: 4
  dangle_parens: true
```

Special configuration will be required for corrosion functions.

I know the project isn't maintained in the recent years, and sometime formatting look weird, but a least it is automatic.

What are you tought about that?

**Comment by ahayzen-kdab** [comment]
Anything which is machine driven and can be added to CI for formatting or linting is good :-)

Is cmake-format the generally accepted solution across CMake projects ? Or are there other options available ? Are you able to point to any large / well know CMake projects using it ?

As being old and not maintained is not necessarily bad if it works, however it is bad if it creates false positives with modern syntax or is hard to get through distribution mechanisms.

And the same would be said about `cmake-lint`.

Probably best way is to add it to CI for Linux, either like `reuse` via pip or via apt and see how it reacts / what the difference in the CMake files is.

**Reply by OlivierLDff** [reply]
I'm using it for all of my project and it work really well.
It's easily extensible to new function in the .cmake-format.

You can see usage in https://github.com/cpm-cmake/CPM.cmake. CI is at https://github.com/cpm-cmake/CPM.cmake/blob/master/.github/workflows/style.yaml

Installation is a simple `pip install cmakelang`. Then it behave like clang-format.

For integration you have vscode, clion and sublime text plugin. I guess for nvim there is ongoing effort in there : https://github.com/regen100/cmake-language-server

This project give cmake target to run clang format & cmake format : https://github.com/TheLartians/Format.cmake

**Reply by ahayzen-kdab** [reply]
OK :-)

Also generally i prefer avoiding having custom options on formatters / linters, and eg just picking defaults or a profile / style, as this avoids endless arguments over which options to enable :-)

So if the default is `tab_size: 2` feel free to have an individual commit that changes all the CMake code to tab size 2. Then the same with dangling_parens, if the default is false and this is accepted generally, do a commit to change the code for that as well before the `.cmake-format` file is introduced.

We probably don't (?) need that extra project to give a cmake target for clang-format and cmake format together i guess, just a ctest for cmake-format. Long term probably something more down the route of `pre-commit` would be the way.

**Reply by OlivierLDff** [reply]
tab_size: 4 & dangling_parens are what match the best of your style
tab_size: 2 & dangling_parens is what is the default in every project I saw.

But maybe I'm biased because I mostly saw project based of https://github.com/TheLartians/ModernCppStarter/blob/master/.cmake-format

**Reply by ahayzen-kdab** [reply]
So then probably start with the following and leave all the other options as default and see what it looks like :-)

```yaml
format:
  dangle_parens: true
```

Which then probably means just the tab size as the biggest mass change.

**Reply by OlivierLDff** [reply]
I will also need to add `# ~~~` around copyright, for them to stay formatted as they are.

**Reply by ahayzen-kdab** [reply]
To stop the first line being wrapped, like we need to with clang-format ? Or does it mess with other lines too ?

As the other route is to move them into a `CMakeLists.txt.license` file that is then separate and not formatted by the formatter :-)

**Reply by OlivierLDff** [reply]
Having everything in one file always tend to be easier.

Header of the file will just look like:
```cmake
# ~~~
# SPDX-FileCopyrightText: 2021 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
# SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
# SPDX-FileContributor: Gerhard de Clercq <gerhard.declercq@kdab.com>
# SPDX-FileContributor: Be <be.0@gmx.com>
#
# SPDX-License-Identifier: MIT OR Apache-2.0
# ~~~
```

Do you have any script applying clang format to every file? Other than the ci check and `clang_format_check.sh` ?

**Reply by ahayzen-kdab** [reply]
Nope, i just press format in the IDE :sweat_smile:  `pre-commit` would be the long term route to do this all properly if that could be hooked into our thing to find the C++ files etc as well.

**Reply by ahayzen-kdab** [reply]
Also moving to `pre-commit` would help with the differing versions of clang-format etc as well as IIRC you can pin it to a version

**Reply by OlivierLDff** [reply]
I've never had to use pre-commit yet as I'm also used to everything being formatted correctly when I save in my IDE. I will see how to do that.
I feel that cmake needs some refactoring, especially around add_test as there is a lot of duplicated code. I'm seeing that while doing my PR to fix ctest on windows.

**Reply by ahayzen-kdab** [reply]
Feel free to split out general changes into other pull requests while you battle fixing the Windows specific issues :-)

**Reply by OlivierLDff** [reply]
I'll first fix the windows problem, then do some general refactor if I have time. Fixing the windows issue force me to go thru every CMakeLists and get good understanding of the project as a whole ;)
